### PR TITLE
Introduce hypothesis testing to BC tests

### PR DESCRIPTION
**Pull request by ernestum**
## Description

This PR introduces hypothesis tests for BC and does some general cleanup such as introducing a GIVEN, WHEN, THEN structure where appropriate  and renaming tests to make it clear what they test.

## Testing

Na, I did not test the tests ...

### CODE REVIEW

File: src/imitation/testing/expert_trajectories.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+from imitation.util import util
+
+
+def generate_expert_trajectories(
+    env_id: str,
+    num_trajectories: int,
+) -> Sequence[types.TrajectoryWithRew]:
+    """Generate expert trajectories for the given environment.
+
+    Note: will just pull a pretrained policy from the huggingface model hub.
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:26]:
```suggestion
    Note: will just pull a pretrained policy from the Hugging Face model hub.
```

File: src/imitation/testing/expert_trajectories.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    Returns:
+        A list of trajectories with rewards.
+    """
+    env = util.make_vec_env(
+        env_id,
+        post_wrappers=[lambda e, _: wrappers.RolloutInfoWrapper(e)],
+    )
+    trajectories = rollout.rollout(
+        serialize.load_policy("ppo-huggingface", env, env_name=env_id),
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:40]: Nit but I'd rather introduce an intermediate variable `policy` to hold this, although only one line it's doing something non-trivial here.

File: src/imitation/testing/expert_trajectories.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        env,
+        rollout.make_sample_until(min_episodes=num_trajectories),
+    )
+    env.close()
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:44]: Consider putting this in a `try:`, `finally:` block?

File: src/imitation/testing/expert_trajectories.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    Args:
+        cache_path: A path to the folder to be used as cache for the expert
+            trajectories.
+        env_id: The environment to generate trajectories for.
+        num_trajectories: The number of trajectories to generate.
+
+    Returns:
+        A list of trajectories with rewards.
+    """
+    environment_cache_path = pathlib.Path(cache_path) / hfsb3.EnvironmentName(env_id)
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:64]: The `EnvironmentName` is doing some path normalization?

File: src/imitation/testing/expert_trajectories.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    env_id: str,
+    num_trajectories: int,
+) -> Sequence[types.TrajectoryWithRew]:
+    """Generate or load expert trajectories from cache.
+
+    Args:
+        cache_path: A path to the folder to be used as cache for the expert
+            trajectories.
+        env_id: The environment to generate trajectories for.
+        num_trajectories: The number of trajectories to generate.
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:59]: Nit: in `generate_expert_trajectories` you describe it as "number of episodes" not "number of trajectories". Both seem OK but should probably be consistent.

File: src/imitation/testing/expert_trajectories.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    environment_cache_path = pathlib.Path(cache_path) / hfsb3.EnvironmentName(env_id)
+    environment_cache_path.mkdir(parents=True, exist_ok=True)
+
+    trajectories_path = environment_cache_path / "rollout.npz"
+
+    with FileLock(environment_cache_path / "rollout.npz.lock"):
+        try:
+            trajectories = types.load_with_rewards(trajectories_path)
+        except (FileNotFoundError, pickle.PickleError):  # pragma: no cover
+            warnings.warn(
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:73]: Nitpick, but this error message seems a bit obscure if it's just a `FileNotFoundError`, which will always happen the first time it's called. Maybe have different warning messages for FileNotFound vs pickle?

File: src/imitation/testing/expert_trajectories.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                "Recomputing expert trajectories due to the following error when "
+                "trying to load them:\n" + traceback.format_exc(),
+            )
+            trajectories = generate_expert_trajectories(env_id, num_trajectories)
+            types.save(trajectories_path, trajectories)
+
+    if len(trajectories) >= num_trajectories:
+        return trajectories[:num_trajectories]
+    else:
+        # If it is not enough, just throw clear the cache and generate more.
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:83]: "just clear the cache" would also be OK
```suggestion
        # If it is not enough, just throw away the cache and generate more.
```

File: src/imitation/testing/expert_trajectories.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        trajectories_path.unlink()
+        return lazy_generate_expert_trajectories(cache_path, env_id, num_trajectories)
+
+
+def make_expert_transition_loader(
+    cache_dir: Path,
+    batch_size,
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:90]:
```suggestion
    batch_size: int,
```

File: src/imitation/testing/expert_trajectories.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    expert_data_type,
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:91]:
```suggestion
    expert_data_type: str,
```

File: src/imitation/testing/expert_trajectories.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    env_name: str,
+    num_trajectories: int = 1,
+):
+    """Creates different kinds of pytorch data loaders for expert transitions.
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:95]:
```suggestion
    """Creates different kinds of PyTorch data loaders for expert transitions.
```

File: src/imitation/testing/expert_trajectories.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    )
+    transitions = rollout.flatten_trajectories(trajectories)
+
+    if len(transitions) < batch_size:
+        # If we have less transitions than the batch size, we estimate the trajectory
+        # length and generate enough trajectories to fill the batch size.
+        trajectory_length = len(transitions) // len(trajectories)
+        min_required_trajectories = math.ceil(batch_size / trajectory_length)
+        transitions = rollout.flatten_trajectories(
+            lazy_generate_expert_trajectories(
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:124]: This might return data that overlaps with `trajectories` -- is that OK? (Probably?)

File: src/imitation/testing/expert_trajectories.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                        acts=self.trans.acts[start:end],
+                    )
+                    d = {k: util.safe_to_tensor(v) for k, v in d.items()}
+                    yield d
+
+        return DucktypedDataset(transitions, batch_size)
+    elif expert_data_type == "transitions":
+        return transitions
+    else:  # pragma: no cover
+        raise ValueError(expert_data_type)
```
Comment (AdamGleave) [inline src/imitation/testing/expert_trajectories.py:167]:
```suggestion
        raise ValueError(f"Unexpected data type '{expert_data_type}'")
```

**Review by AdamGleave — COMMENTED** [review]
Thanks for implementing this! The tests are much easier to understand, and I like the use of hypothesis, should give us more comprehensive coverage. I've mostly just left some minor comments.

I'm a bit unsure about adding some of the additions `imitation.testing`. It seems best to reserve this for generally useful methods, that will either be used across multiple files in our test suite (and doesn't make sense to share as a fixture in `conftest`) or by users of the library. 

I think the `expert_trajectories.py` largely satisfies that, though maybe it could live in `conftest.py` instead as it seems to only be used via some fixtures right now. But `failing_data_loader.py` as far as I can tell is used in just a single test in `test_bc.py` -- why not just define it alongside the test?

**Comment by AdamGleave** [comment]
Also one test is failing on Mac OS (maybe flaky? I don't see anything Mac OS specific). And type checking failing --- I think pytype may be confused by the presence of two `conftest.py` files, adding `__init__.py` to `tests/algorithms` may help with this.

File: src/imitation/testing/expert_trajectories.py (hunk 4 above)
Comment (ernestum) [reply src/imitation/testing/expert_trajectories.py:64]: Yes, I introduced it [here](https://github.com/huggingface/huggingface_sb3/pull/13). It basically just removes slashes from environment names and ensures that environment names are normalized in an uniform way.

File: src/imitation/testing/expert_trajectories.py (hunk 3 above)
Comment (ernestum) [reply src/imitation/testing/expert_trajectories.py:44]: Yes good idea!

File: src/imitation/testing/expert_trajectories.py (hunk 5 above)
Comment (ernestum) [reply src/imitation/testing/expert_trajectories.py:61]: Thanks, fixed!

File: src/imitation/testing/expert_trajectories.py (hunk 6 above)
Comment (ernestum) [reply src/imitation/testing/expert_trajectories.py:75]: Good point! Users would probably be clueless.
I tried to give better warnings. Not sure if I like the implementation. What do you think?

**Comment by ernestum** [comment]
> Also one test is failing on Mac OS (maybe flaky? I don't see anything Mac OS specific)

Me neither. It definitively is one of the flaky tests. Should probably be investigated in a separate PR.

**Comment by ernestum** [comment]
> And type checking failing

Yep, thanks for the hint. That fixed it.

**Comment by ernestum** [comment]
> I'm a bit unsure about adding some of the additions `imitation.testing`. It seems best to reserve this for generally useful methods, that will either be used across multiple files in our test suite (and doesn't make sense to share as a fixture in `conftest`) or by users of the library.

I agree. I was unsure where to place all the helpers too. In the end I decided to move them out of the way to improve readability of the tests. Placing them into `conftest.py` was somewhat weird because of the special treatment it gets. I assumed it was reserved for fixtures? Where would you recommend to put it? Maybe just moving the failing data loader back to the test is reasonable ...

File: src/imitation/testing/expert_trajectories.py (hunk 11 above)
Comment (AdamGleave) [reply src/imitation/testing/expert_trajectories.py:131]: Yeah, I think this is probably fine as-is, the solutions I can think of would add more complexity than they're worth.

File: src/imitation/testing/expert_trajectories.py (hunk 6 above)
Comment (AdamGleave) [reply src/imitation/testing/expert_trajectories.py:80]: Seems fine. I might still include the traceback just for extra info.

**Comment by AdamGleave** [comment]
> Maybe just moving the failing data loader back to the test is reasonable ...

I think this is the least bad option. I don't think it distracts from the tests that much, especially if you add a part of the file designated by a comment header for helper methods.

**Comment by ernestum** [comment]
I moved the data loader back to `test_bc.py`.

The test is only flaky on the MacOS runner. So this will take some more time to investigate since I don't have a MacOS system right now.

File: tests/algorithms/test_bc.py [hunk 13]
```diff
@@ -259,23 +287,41 @@ def test_that_bc_raises_error_when_data_loader_is_empty(
...
     trans = rollout.flatten_trajectories(cartpole_expert_trajectories)
     dummy_yield_value = dataclasses.asdict(trans[:batch_size])
 
-    bad_data_loader = failing_data_loader.DataLoaderThatFailsOnNthIter(
-        dummy_yield_value=dummy_yield_value,
-        # add 1 as BC uses up an iteration from getting the first element
-        # for type checking
-        no_yield_after_iter=no_yield_after_iter + 1,
-    )
+    class DataLoaderThatFailsOnNthIter:
```
Comment (AdamGleave) [inline tests/algorithms/test_bc.py:295]: Our style guide favors classes at top-level except if they're closing over variables: https://google.github.io/styleguide/pyguide.html#26-nestedlocalinner-classes-and-functions

Can you move it? You can prefix it with an underscore to indicate it's private though I don't think it matters given this is in a test file.

**Review by AdamGleave — APPROVED** [review]
LGTM -- one minor style suggestion, but no need for me to re-review

File: tests/algorithms/test_bc.py (hunk 13 above)
Comment (ernestum) [reply tests/algorithms/test_bc.py:294]: Thanks for the hint. Turns out I was closing over the dummy yield value.

I updated the implementation to also close over `no_yield_after_iter`. This way I could simplify the  `DataLoaderThatFailsOnNthIter` even more.

I also used existing fixtures to remove some more setup code and added some code to ensure that the training loop actually executed `no_yield_after_iter` many times. This implicitly tests the `DataLoaderThatFailsOnNthIter` to address the concerns in the style guide.

Comment (AdamGleave) [reply tests/algorithms/test_bc.py:294]: Ah, I missed it was closing over `dummy_yield_value`. I agree keeping it an inner class makes sense, and this is an elegant solution that simplifies the code :)

File: tests/algorithms/test_bc.py [hunk 14]
```diff
@@ -268,60 +268,49 @@ def test_that_wrong_training_duration_specification_raises_error(
...
-        def __init__(self, no_yield_after_iter: int):
-            """Builds dummy data loader.
-
-            Args:
-                no_yield_after_iter: `__iter__` will raise `StopIteration` after
-                    this many calls.
-            """
-            self.iter_count = 0
-            self.no_yield_after_iter = no_yield_after_iter
+        iter_count = 0
```
Comment (AdamGleave) [inline tests/algorithms/test_bc.py:297]: Nit: Using a class rather than instance variable here is a bit odd. Doesn't matter right now as we only ever have a single instance of the class. But it could be a gotcha if we change the test later, so I lean slightly towards keeping it in `__init__` even though it's more code?

**Review by AdamGleave — APPROVED** [review]
LGTM, one minor suggestion to make iter_count instance rather than class variable.

File: tests/algorithms/test_bc.py (hunk 14 above)
Comment (ernestum) [reply tests/algorithms/test_bc.py]: Fixed this.
